### PR TITLE
Add Thycotic Secret Server plugin #21

### PR DESCRIPTION
**Pull request by jifox**
Implement: Issue #21 

I've implemented a Secret Provider for Thycotic Secret Server https://thycotic.com.

In my Development Environment is is working with an on premise server as expected. I'm able to synchronize "Git Repositories" with secrets from Thycotic provider.

Actually I have problems with integrating the new provider with the development environment  used in this repository.

### Error: Invoke command

```bash
$  invoke -e build --no-cache
Building Nautobot with Python 3.7...
Running docker-compose command "build --no-cache"
docker-compose --project-name nautobot_secrets_providers --project-directory "development" -f "development/docker-compose.requirements.yml" -f "development/docker-compose.base.yml" -f "development/docker-compose.dev.yml" build --no-cache
vault uses an image, skipping
redis uses an image, skipping
postgres uses an image, skipping
worker uses an image, skipping
Building nautobot
Sending build context to Docker daemon  757.2kB
Step 1/14 : ARG NAUTOBOT_VER="1.2.1"
Step 2/14 : ARG PYTHON_VER=3.8
Step 3/14 : FROM ghcr.io/nautobot/nautobot-dev:${NAUTOBOT_VER}-py${PYTHON_VER}
 ---> d9c3a2044285
Step 4/14 : ENV prometheus_multiproc_dir=/prom_cache
 ---> Running in 6a2e8ea80e2d
Removing intermediate container 6a2e8ea80e2d
 ---> deaa0ddd6296
Step 5/14 : ARG NAUTOBOT_ROOT=/opt/nautobot
 ---> Running in e7e24793c448
Removing intermediate container e7e24793c448
 ---> 397bb47b794b
Step 6/14 : ENV NAUTOBOT_ROOT ${NAUTOBOT_ROOT}
 ---> Running in 47dc421a2620
Removing intermediate container 47dc421a2620
 ---> 2e1549672f45
Step 7/14 : WORKDIR $NAUTOBOT_ROOT
 ---> Running in d5c4b6cbdb39
Removing intermediate container d5c4b6cbdb39
 ---> ced99e8597f6
Step 8/14 : RUN poetry config virtualenvs.create false     && poetry config installer.parallel false
 ---> Running in edb205676591
Removing intermediate container edb205676591
 ---> fe2aa00b2327
Step 9/14 : WORKDIR /source
 ---> Running in b5a6ebeba8c5
Removing intermediate container b5a6ebeba8c5
 ---> a548c4e01e11
Step 10/14 : COPY poetry.lock pyproject.toml /source/
 ---> 6b644f2b2185
Step 11/14 : RUN poetry install --no-interaction --no-ansi --no-root
 ---> Running in 530c39f4438c
Skipping virtualenv creation, as specified in config file.
Installing dependencies from lock file
Warning: The lock file is not up to date with the latest changes in pyproject.toml. You may be getting outdated dependencies. Run update to update them.

  SolverProblemError

  Because nautobot-secrets-providers depends on python-tss-sdk (>=1.0.0,<2.0.0) which doesn't match any versions, version solving failed.

  at ~/.local/share/pypoetry/venv/lib/python3.7/site-packages/poetry/puzzle/solver.py:241 in _solve
      237│             packages = result.packages
      238│         except OverrideNeeded as e:
      239│             return self.solve_in_compatibility_mode(e.overrides, use_latest=use_latest)
      240│         except SolveFailure as e:
    → 241│             raise SolverProblemError(e)
      242│ 
      243│         results = dict(
      244│             depth_first_search(
      245│                 PackageNode(self._package, packages), aggregate_package_nodes
The command '/bin/sh -c poetry install --no-interaction --no-ansi --no-root' returned a non-zero code: 1
Service 'nautobot' failed to build : Build failed
```

> Because nautobot-secrets-providers depends on python-tss-sdk (>=1.0.0,<2.0.0) which doesn't match any versions, version solving failed.

When checking with pip I see the packages "1.0.0" and "1.0.1":

```bash
$ pip install python-tss-sdk==
ERROR: Could not find a version that satisfies the requirement python-tss-sdk== (from versions: 0.0.1, 0.0.2, 0.0.3, 0.0.4, 0.0.5, 0.0.6, 1.0.0, 1.0.1)
ERROR: No matching distribution found for python-tss-sdk==
```

### invoke.yml

```yml
---
nautobot_secrets_providers:
  project_name: "nautobot_secrets_providers"
  nautobot_ver: "1.2.0"
  local: false
  python_ver: "3.7"
  compose_dir: "development"
  compose_files:
    - "docker-compose.requirements.yml"
    - "docker-compose.base.yml"
    - "docker-compose.dev.yml"
```

### pyproject.toml

```toml
[tool.poetry]
name = "nautobot-secrets-providers"
version = "1.1.0"
description = "Nautobot Secrets Providers Plugin."
authors = ["Network to Code, LLC <opensource@networktocode.com>"]
license = "Apache-2.0"
readme = "README.md"
homepage = "https://github.com/nautobot/nautobot-plugin-secrets-providers"
repository = "https://github.com/nautobot/nautobot-plugin-secrets-providers"
keywords = ["nautobot", "nautobot-plugin"]
include = [
    "LICENSE",
    "README.md",
]
packages = [
    { include = "nautobot_secrets_providers" },
]

[tool.poetry.dependencies]
python = "^3.6.2"
# Pin this to the lowest supported version of Nautobot
nautobot = "~1.2.1"

#
# Optional dependencies installed as extras
# 

# AWS SDK for Python (always keep in sync w/ dev dependency)
boto3 = { version = "~1.19.5", optional = true }
# HashiCorp Vault API Client (always keep in sync w/ dev dependency)
hvac = { version = "~0.11.0", optional = true }
# Thycotic Secret Server SDK (always keep in sync w/ dev dependency)
python-tss-sdk = { version = ">=1.0.0,<2.0.0", optional = true }

[tool.poetry.dev-dependencies]
bandit = "*"
black = "*"
boto3 = "~1.19.5"  # (always keep in sync w/ optional dependency)
coverage = "*"
django-debug-toolbar = "*"
flake8 = "*"
hvac = "~0.11.0"  # (always keep in sync w/ optional dependency)
invoke = "*"
moto = "~2.2.11"
mkdocs = "*"
pydocstyle = "*"
pylint = "*"
pylint-django = "*"
python-tss-sdk = ">=1.0.0,<2.0.0" # (always keep in sync w/ optional dependency)
requests-mock = "~1.9.3"
yamllint = "*"
nautobot = "^1.2.0"

[tool.poetry.extras]
all = ["boto3", "hvac", "thycotic"]
aws = ["boto3"]
hashicorp = ["hvac"]
nautobot = ["nautobot"]
thycotic = ["thycotic"] 

[tool.black]
line-length = 120
target-version = ['py37']
include = '\.pyi?$'
exclude = '''
(
  /(
      \.eggs         # exclude a few common directories in the
    | \.git          # root of the project
    | \.hg
    | \.mypy_cache
    | \.tox
    | \.venv
    | _build
    | buck-out
    | build
    | dist
  )/
  | settings.py     # This is where you define files that should not be stylized by black
                     # the root of the project
)
'''

[tool.pylint.master]
# Include the pylint_django plugin to avoid spurious warnings about Django patterns
load-plugins="pylint_django"

[tool.pylint.basic]
# No docstrings required for private methods (Pylint default), or for test_ functions, or for inner Meta classes.
no-docstring-rgx="^(_|test_|Meta$)"
# Don't lint the test config
ignore = ["nautobot_config.py"]

[tool.pylint.messages_control]
# Line length is enforced by Black, so pylint doesn't need to check it.
# Pylint and Black disagree about how to format multi-line arrays; Black wins.
disable = """,
    line-too-long,
    bad-continuation,
    """

[tool.pylint.miscellaneous]
# Don't flag TODO as a failure, let us commit with things that still need to be done in the code
notes = """,
    FIXME,
    XXX,
    """

[build-system]
requires = ["poetry_core>=1.0.0"]
build-backend = "poetry.core.masonry.api"

[tool.pytest.ini_options]
testpaths = [
    "tests"
]
addopts = "-vv --doctest-modules"
```

Has anyone a clue whats not working in this development environment?

### CODE REVIEW

**Comment by glennmatthews** [comment]
`poetry add --optional 'python-tss-sdk>=1.0.0,<2.0'` provides a bit more detail as to the dependency issue for me:

```
  SolverProblemError

  The current project's Python requirement (3.9.10) is not compatible with some of the required packages Python requirement:
    - dataclasses requires Python >=3.6, <3.7, so it will not be satisfied for Python 3.9.10
  
  Because no versions of python-tss-sdk match >=1.0.0,<1.0.1 || >1.0.1,<2.0
   and python-tss-sdk (1.0.1) depends on dataclasses (>=0.6), python-tss-sdk (>=1.0.0,<2.0) requires dataclasses (>=0.6).
  Because dataclasses (0.8) requires Python >=3.6, <3.7
   and no versions of dataclasses match >=0.6,<0.8 || >0.8, dataclasses is forbidden.
  Thus, python-tss-sdk is forbidden.
  So, because nautobot-secrets-providers depends on python-tss-sdk (>=1.0.0,<2.0), version solving failed.
```

Basically it looks like `python-tss-sdk` states that it's unconditionally dependent on the `dataclasses` package, which is a backport of Python 3.7+ built-in functionality to Python 3.6 and as such isn't intended to be compatible with any later Python version. This seems to be a bug in the packaging of `python-tss-sdk` to me. 

I was able to seemingly work around this issue by explicitly pinning to an older version of `dataclasses` that, while still incompatible with later Python versions according to https://github.com/ericvsmith/dataclasses#compatibility, doesn't explicitly block installation under later Python versions:

```
poetry add --optional dataclasses==0.6
poetry add --optional 'python-tss-sdk>=1.0.0,<2.0'
```

But definitely a preferable approach would be to get `python-tss-sdk` to fix their dependency declaration. It looks like https://github.com/thycotic/python-tss-sdk/issues/32 reported this issue previously but was incorrectly closed as a cosmetic-only issue.

File: nautobot_secrets_providers/providers/thycotic.py [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+                    (
+                        "Thycotic Secret Server is not configured properly! "
+                        "Trusted certificates file not found: "
+                        "Environment variable 'REQUESTS_CA_BUNDLE': "
+                        f"{ca_bundle_path}."
+                    ),
+                )
+            from_env = os.getenv("REQUESTS_CA_BUNDLE", "")
+            if from_env != ca_bundle_path:
+                os.environ["REQUESTS_CA_BUNDLE"] = ca_bundle_path
```
Comment (glennmatthews) [inline nautobot_secrets_providers/providers/thycotic.py:139]: This feels like it could have unintended consequences to usage of `requests` elsewhere in this plugin or any other Nautobot plugin. Maybe we should have a `try... finally` block that restores the original value of `$REQUESTS_CA_BUNDLE` when this function is done?

**Review by glennmatthews — COMMENTED** [review]
Nice work!

File: nautobot_secrets_providers/providers/thycotic.py (hunk 1 above)
Comment (jifox) [reply nautobot_secrets_providers/providers/thycotic.py]: I commited the change now

File: pyproject.toml [hunk 2]
```diff
@@ -29,6 +29,9 @@ nautobot = "~1.2.1"
 boto3 = { version = "~1.19.5", optional = true }
 # HashiCorp Vault API Client (always keep in sync w/ dev dependency)
 hvac = { version = "~0.11.0", optional = true }
+# Thycotic Secret Server (always keep in sync w/ dev dependency)
+python-tss-sdk = {version = ">=1.0.1,<2.0", optional = true}
```
Comment (glennmatthews) [inline pyproject.toml:33]: It looks like the next version of python-tss-sdk (which will remove the `dataclasses` dependency) is also going to drop support for python 3.6 (understandably) but as Nautobot 1.2.x still supports python 3.6, it might be desirable to add a `python = ">=3.7"` to this dependency, and document this restriction in the README.

Does this PR need to be blocked until a new release of python-tss-sdk is published?

File: pyproject.toml [hunk 3]
```diff
@@ -41,17 +44,21 @@ hvac = "~0.11.0"  # (always keep in sync w/ optional dependency)
 invoke = "*"
 moto = "~2.2.11"
 mkdocs = "*"
+# Pin this to the lowest supported version of Nautobot
+nautobot = "~1.2.1"
```
Comment (glennmatthews) [inline pyproject.toml:48]: This shouldn't need to be added to `dev-dependencies` as it's already a core dependency of the project (line 22 above).

Comment (jifox) [reply pyproject.toml]: Updated pyproject and fixed pydocstyle, flake8

File: pyproject.toml (hunk 2 above)
Comment (EndlessTrax) [reply pyproject.toml:33]: I just released the latest version of the `python-tss-sdk` (1.1.0) to PyPI. I hope this helps the PR author to move this along. Great work @jifox.

File: nautobot_secrets_providers/providers/__init__.py [hunk 4]
```diff
@@ -2,5 +2,10 @@
...
 from .aws import AWSSecretsManagerSecretsProvider
 from .hashicorp import HashiCorpVaultSecretsProvider
+from .thycotic import ThycoticSecretServerSecretsProvider
 
-__all__ = (AWSSecretsManagerSecretsProvider, HashiCorpVaultSecretsProvider)  # pylint: disable=invalid-all-object
+__all__ = (  # type: ignore
+    AWSSecretsManagerSecretsProvider,
+    HashiCorpVaultSecretsProvider,
+    ThycoticSecretServerSecretsProvider,
+)
```
[the remark this answers is not in the document]
Comment (jifox) [reply nautobot_secrets_providers/providers/__init__.py]: @glennmatthews Hi, I had to restore the __all__ as objects because in secrets.py the import will use this objects.

__all__ = (  # type: ignore
    AWSSecretsManagerSecretsProvider,
    HashiCorpVaultSecretsProvider,
    ThycoticSecretServerSecretsProvider,
)

**Comment by jifox** [comment]
@glennmatthews As the version was 1.0.0 and new thycotic functionality is added I'll going to increase the minor version to 1.1.0

Where should the Release Notes be documented?

**Comment by glennmatthews** [comment]
You can update the CHANGELOG.md for the new version. Thanks!

**Comment by jifox** [comment]
@glennmatthews I've successfully tested the on premise secret provider. I have no access to Cloud services so I couldn't test it.

File: nautobot_secrets_providers/providers/thycotic.py [hunk 5]
```diff
@@ -0,0 +1,240 @@
...
+    """A secrets provider for Thycotic Secret Server."""
+
+    slug = "thycotic-tss-id"  # type: ignore
+    name = "Thycotic Secret Server by ID"  # type: ignore
+
+    class ParametersForm(BootstrapMixin, forms.Form):
+        """Required parameters for Thycotic Secret Server."""
+
+        secret_id = forms.IntegerField(
+            required=True,
```
Comment (glennmatthews) [inline nautobot_secrets_providers/providers/thycotic.py:198]:
```suggestion
            label="Secret ID",
            required=True,
```

File: pyproject.toml [hunk 6]
```diff
@@ -44,14 +47,17 @@ mkdocs = "*"
...
 yamllint = "*"
+django-debug = "^0.1.1"
 
 [tool.poetry.extras]
-all = ["boto3", "hvac"]
+all = ["boto3", "hvac", "thycotic"]
 aws = ["boto3"]
 hashicorp = ["hvac"]
 nautobot = ["nautobot"]
+thycotic = ["thycotic"]
```
Comment (glennmatthews) [inline pyproject.toml:60]: I believe these should be:

```suggestion
all = ["boto3", "hvac", "python-tss-sdk"]
aws = ["boto3"]
hashicorp = ["hvac"]
nautobot = ["nautobot"]
thycotic = ["python-tss-sdk"]
```

**Comment by jifox** [comment]
@glennmatthews I fixed the errors and manually verified the functionality again by using the Thycotic Secret Server Provider with an on-prem server and synving a git repository with it

image

**Comment by jifox** [comment]
@glennmatthews How to proceed?  Python 3.7 required

**Comment by glennmatthews** [comment]
It should be possible to add the `python-tss-sdk` dependency as conditional on python 3.7+ so that the the plugin as a whole can still be installed under python 3.6, just without support for thycotic. I'm working on making the appropriate changes and testing them. Thanks for the followup!

File: README.md [hunk 7]
```diff
@@ -29,9 +30,17 @@ This plugin supports the following popular secrets backends:
...
+![Screenshot of secret using Thycotic Secret Server by ID](https://raw.githubusercontent.com/nautobot/nautobot-plugin-secrets-providers/develop/docs/images/screenshot05.png "Secret using Thycotic Secret Server by ID")
+
+---
+
+![Screenshot of secret using Thycotic Secret Server by Path](https://raw.githubusercontent.com/nautobot/nautobot-plugin-secrets-providers/develop/docs/images/screenshot06.png "Secret using Thycotic Secret Server by Path")
+
 ## Installation
 
-> Nautobot Secrets Providers is compatible with Nautobot 1.2.0 and higher
+> Nautobot Secrets Providers is compatible with Nautobot 1.2.0 and higher on Python 3.7 and higher. Python 3.6 is not supported.
```
Comment (glennmatthews) [inline README.md:43]:
```suggestion
> Nautobot Secrets Providers is compatible with Nautobot 1.2.0 and higher. Support for Thycotic Secret Server requires Python 3.7 or later.
```

**Comment by glennmatthews** [comment]
Sure would be easier if we *did* drop Python 3.6 support, since so many of our dependencies have already done so. Still working on this.

File: nautobot_secrets_providers/providers/thycotic.py [hunk 8]
```diff
@@ -0,0 +1,247 @@
...
+            cloud_based=thycotic_settings["cloud_based"],
+            domain=thycotic_settings["domain"] if is_valid_domain else None,
+            password=thycotic_settings["password"] if is_valid_password else None,
+            secret_id=secret_id,
+            secret_path=secret_path,
+            secret_selected_value=secret_selected_value,
+            tenant=thycotic_settings["tenant"] if is_valid_tenant else None,
+            token=thycotic_settings["token"] if is_valid_token else None,
+            username=thycotic_settings["username"] if is_valid_username else None,
+            caller_class=cls,
```
Comment (jathanism) [inline nautobot_secrets_providers/providers/thycotic.py:91]: It appears to me that since you're just checking for the presence of a key in `thycotic_settings` for each of these `is_valid_foo` variables, and then passing them using the ternary operator to `ls.query_thycotic_secret_server()` as `get the key from the dict else None`, this could be simplified for readability and maintainability.

For example, instead of:

```python
        is_valid_base_url = "base_url" in thycotic_settings and thycotic_settings["base_url"] != ""
```

You could do:

```python
        base_url = thycotic_settings.get("base_url", None)
        if base_url is None:
            raise exceptions.SecretProviderError(secret, cls, "Thycotic Secret Server is not configured!")
```

When you pass it to `cls.query_thycotic_secret_server()`, instead you could just pass `base_url=base_url`.

For the rest of them, the same is also true. For example for `domain`. This:

```python
        is_valid_domain = "domain" in thycotic_settings and thycotic_settings["domain"] != ""
        # ... 
        domain=thycotic_settings["domain"] if is_valid_domain else None,
```

Could be simplified to:

```python
        domain = thycotic_settings.get("domain", None)
        # ... 
        domain=domain,
```

And then you could repeat this for the remaining arguments you're pulling in from `thycotic_settings`.

File: nautobot_secrets_providers/providers/thycotic.py [hunk 9]
```diff
@@ -0,0 +1,247 @@
...
+                            "Thycotic Secret Server is not configured properly! "
+                            "Trusted certificates file not found: "
+                            "Environment variable 'REQUESTS_CA_BUNDLE': "
+                            f"{ca_bundle_path}."
+                        ),
+                    )
+                if original_env != ca_bundle_path:
+                    os.environ["REQUESTS_CA_BUNDLE"] = ca_bundle_path
+                    must_restore_env = True
+            # Setup thycotic authorizer
```
Comment (jathanism) [inline nautobot_secrets_providers/providers/thycotic.py:144]:
```suggestion
            # Setup Thycotic authorizer
```

File: nautobot_secrets_providers/providers/thycotic.py [hunk 10]
```diff
@@ -0,0 +1,247 @@
...
+        username=None,
+        caller_class=None,
+    ):
+        """Query Thycotic Secret Server."""
+        # Ensure required parameters are set
+        if (
+            token is None and (username is None or password is None)
+        ) or (  # pylint: disable=too-many-boolean-expressions
+            cloud_based and (tenant is None or username is None or password is None)
+        ):
```
Comment (jathanism) [inline nautobot_secrets_providers/providers/thycotic.py:116]: This is really hard to follow and either needs to be simplified (see my suggestion below) or broken into two separate conditions one each for `token` and one for `cloud_based` conditions.

Based on how I'm reading the code, here is my suggested readability improvement. Altho since I haven't tested it, you would need to double check that it satisfies the same intent:
```suggestion
        if any([
            token is None and not any([username, password]),
            cloud_based and not any([tenant, username, password])
        ]):
```

File: nautobot_secrets_providers/providers/thycotic.py [hunk 11]
```diff
@@ -0,0 +1,247 @@
...
+                    os.environ["REQUESTS_CA_BUNDLE"] = ca_bundle_path
+                    must_restore_env = True
+            # Setup thycotic authorizer
+            # Username | Password | Token | Domain | Authorizer
+            #   def    |   def    |   *   |   -    | PasswordGrantAuthorizer
+            #   def    |   def    |   *   |  def   | DomainPasswordGrantAuthorizer
+            #    -     |    -     |  def  |   *    | AccessTokenAuthorizer
+            #   def    |    -     |  def  |   *    | AccessTokenAuthorizer
+            #    -     |   def    |  def  |   *    | AccessTokenAuthorizer
+            if username is not None and password is not None:
```
Comment (jathanism) [inline nautobot_secrets_providers/providers/thycotic.py:151]:
```suggestion
            if all([username, password]):
```

File: nautobot_secrets_providers/providers/thycotic.py [hunk 12]
```diff
@@ -0,0 +1,247 @@
...
+        )
+        secret_selected_value = forms.ChoiceField(
+            label="Return value",
+            required=True,
+            choices=(
+                ("password", "Password"),
+                ("username", "Username"),
+                ("url", "URL"),
+                ("notes", "Notes"),
+            ),
```
Comment (jathanism) [inline nautobot_secrets_providers/providers/thycotic.py:217]: Since you're using this same choice set in both providers, can you please make this into a `ChoiceSet` defined at the module-level and passed to the `ChoiceField.choices` for each?

```python
from nautobot.utilities.choices import ChoiceSet

class ThycoticSecretChoices(ChoiceSet):
    SECRET_PASSWORD = "password"
    SECRET_USERNAME = "username"
    SECRET_URL = "url"
    SECRET_NOTES = "notes"

    CHOICES = (
        (SECRET_PASSWORD, "Password"),
        (SECRET_USERNAME, "Username"),
        (SECRET_URL, "URL"),
        (SECRET_NOTES, "Notes"),
    )

# And then on the field for each provider's `ParametersForm`:
            choices=ThycoticSecretChoices,

**Review by jathanism — CHANGES_REQUESTED** [review]
Thank you so much for the contribution! This is looking really great, however, I do have some readability, style, and long-term maintainability/reusability changes I would like to see. I provided suggestions to simplify it for you. Thanks again!

File: nautobot_secrets_providers/providers/thycotic.py (hunk 8 above)
Comment (jifox) [reply nautobot_secrets_providers/providers/thycotic.py:91]:
```
is_valid_base_url = "base_url" in thycotic_settings and thycotic_settings["base_url"] != ""
if not is_valid_base_url:
    ....
```

I changed the documentation to use `None` as default value instead of `""`.

```
PLUGINS_CONFIG = {
    "nautobot_secrets_providers": {
        "thycotic": {  # https://github.com/thycotic/python-tss-sdk
            "base_url": os.getenv("SECRET_SERVER_BASE_URL", None),
            "ca_bundle_path": os.getenv("REQUESTS_CA_BUNDLE", None),
            ...
```

This will allow to use the suggested code

File: nautobot_secrets_providers/providers/thycotic.py (hunk 10 above)
Comment (jifox) [reply nautobot_secrets_providers/providers/thycotic.py:116]: This will replace the abve if

```
  if any([
    token is None and not all([username, password]),
    cloud_based and not all([tenant, username, password])
  ]):
```

**Comment by jifox** [comment]
@jathanism  Thanks for detailed review.

- All changes are committed
- `invoke tests` leads to no errors
- Tested with Secret Server On-Prem

**Comment by jifox** [comment]
@jathanism @glennmatthews Hello, ist there anything missing, whats blocking the merge?

**Comment by bryanculver** [comment]
Apologies for the delay @jifox. We'll get this re-reviewed in the next day or two to get merged. Thank you for all of the prompt response to feedback!

**Review by jathanism — APPROVED** [review]
:shipit:

Thanks again @jifox !

**Comment by EndlessTrax** [comment]
Love seeing this get merged. Well done @jifox 😁
